### DISCUSSION TITLE
Window customization with rust frontend

### DISCUSSION THREAD

**Discussion by StepanZagray** [tauri-apps/tauri#10279] [Q&A] [answered]
In [docs](https://v2.tauri.app/plugin/window-customization/) there is only javascript example for implementing functionality for titlebar buttons. After I set all permissions and css, I created `<script>` in `index.html` and paste code there. Like this: 

```html
    <body>
        <div data-tauri-drag-region class="titlebar">
            <div class="titlebar-button" id="titlebar-minimize">
                <img
                    src="https://api.iconify.design/mdi:window-minimize.svg"
                    alt="minimize"
                />
            </div>
            <div class="titlebar-button" id="titlebar-maximize">
                <img
                    src="https://api.iconify.design/mdi:window-maximize.svg"
                    alt="maximize"
                />
            </div>
            <div class="titlebar-button" id="titlebar-close">
                <img
                    src="https://api.iconify.design/mdi:close.svg"
                    alt="close"
                />
            </div>
        </div>
        <script>
            const { Window } = window.__TAURI__.window

            const appWindow = new Window('main')

            document
                .getElementById('titlebar-minimize')
                ?.addEventListener('click', () => appWindow.minimize())
            document
                .getElementById('titlebar-maximize')
                ?.addEventListener('click', () => appWindow.toggleMaximize())
            document
                .getElementById('titlebar-close')
                ?.addEventListener('click', () => appWindow.close())
        </script>
    </body>
```

After compiling, buttons are not working and all leptos frontend is gone. In web console there are these errors:

```
Uncaught Error: 
`unwrap_throw` failed (C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\leptos_dom-0.6.12\src\lib.rs:1026:71)
```

and

```
POST http://ipc.localhost/plugin%3Awindow%7Ccreate 400 (Bad Request)
```

Also depending on what button I click console gives these errors:

```
POST http://ipc.localhost/plugin%3Awindow%7Ctoggle_maximize 400 (Bad Request)
```

```
Uncaught (in promise)
window.toggle_maximize not allowed. Permissions associated with this command: window:allow-toggle-maximize
```

How to do window customization in Tauri Leptos app correctly?

**Comment by FabianLars** [comment]
Sounds like you missed the permission section https://v2.tauri.app/plugin/window-customization/#permissions it applies no matter what frontend stack you're using.

Edit: just saw you said you set those, apologies. But still, that's what the error is saying 🤔

**Reply by StepanZagray** [reply]
Yea I actually only wrote "window:allow-start-dragging" in permissions, my bad😅. But I still have first two errors:
```
Uncaught Error: 
`unwrap_throw` failed (C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\leptos_dom-0.6.12\src\lib.rs:1026:71)
```

and

```
POST http://ipc.localhost/plugin%3Awindow%7Ccreate 400 (Bad Request)
```

**Reply by StepanZagray** [reply]
I used tauri leptos template generated by ```cargo create-tauri-app --beta```. I didn't change anything in configs besides what is written in docs about window customization. From error I assume its related to leptos, but this error occurs only if I write <script> tag into index.html. Without it there is no errors

**Reply by StepanZagray** [reply]
To make sure sure everything is right I created new tauri leptos project, did steps described in docs - same errors. I am interested do you have same errors or is it on my machine only?

**Reply by FabianLars** [reply]
Just got back from the weekend, i'll take a look in a few!

**Reply by StepanZagray** [reply]
The problem is in those lines: 
```
const { Window } = window.__TAURI__.window
const appWindow = new Window('main')
```
There is no errors and frontend renders if I comment these lines

**Reply by FabianLars** [reply]
Does adding `window:allow-create` make it work? I still didn't try it myself but the error suggests that's the issue and according to tauri's code it's not part of `window:default` which means the guide is incomplete/incorrect.

**Reply by StepanZagray** [reply]
I am actually now getting 
```
Failed to load resource: the server responded with a status of 400 (Bad Request)   ipc.localhost/plugin%3Awindow%7Ccreate:1          
```
no matter if there is `window:allow-create ` or not in permissions
and there is still this error I mentioned in the question:
```
 Uncaught Error: 
`unwrap_throw` failed (C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\leptos_dom-0.6.12\src\lib.rs:1026:71)
```

**Reply by StepanZagray** [reply]
And yes, by the way, in docs of other plugins it says that you need to download plugin for example `cargo tauri add os`. Maybe I need also do the same with window customization?

**Reply by FabianLars** [reply]
The guide makes no sense here tbh. Replace the new Window stuff with this and it should work (it does for me then)
```html
      const { getCurrentWindow } = window.__TAURI__.window;

      const appWindow = getCurrentWindow();
```

`new Window` will always try to create a _new_ window even if the label matches the current one.

**Reply by StepanZagray** [reply]
I got
```
Uncaught TypeError: getCurrentWindow is not a function
```

**Reply by FabianLars** [reply, marked as answer]
Ah then you're not on the latest version. Try just `getCurrent` then

**Reply by StepanZagray** [reply]
Thanks, everything works good on newest version. I will just use newer version
